src/rabbit_mqtt_collector.erl

```diff
@@ -11,91 +11,33 @@
+    run_ra_command({register, ClientId, Pid}).
 
 unregister(ClientId, Pid) ->
-    gen_server:call(rabbit_mqtt_collector, {unregister, ClientId, Pid}, infinity).
+    run_ra_command({unregister, ClientId, Pid}).
 
 list() ->
-    gen_server:call(rabbit_mqtt_collector, list).
-
-%%----------------------------------------------------------------------------
-
-init([]) ->
-    {ok, #state{client_ids = #{}}}. % clientid -> {pid, monitor}
+     run_ra_command(list).
 
-%%--------------------------------------------------------------------------
+leave(NodeBin) ->
+    Node = binary_to_atom(NodeBin, utf8),
+    run_ra_command({leave, Node}),
+    mqtt_node:leave(Node).
 
-handle_call({register, ClientId, Pid}, _From,
-            State = #state{client_ids = Ids}) ->
-    Ids1 = case maps:find(ClientId, Ids) of
-               {ok, {OldPid, MRef}} when Pid =/= OldPid ->
-                   catch gen_server2:cast(OldPid, duplicate_id),
-                   erlang:demonitor(MRef),
-                   maps:remove(ClientId, Ids);
-               error ->
-                   Ids
-           end,
-    Ids2 = maps:put(ClientId, {Pid, erlang:monitor(process, Pid)}, Ids1),
-    {reply, ok, State#state{client_ids = Ids2}};
-
-handle_call({unregister, ClientId, Pid}, _From, State = #state{client_ids = Ids}) ->
-    {Reply, Ids1} = case maps:find(ClientId, Ids) of
-                        {ok, {Pid, MRef}} -> erlang:demonitor(MRef),
-                                             {ok, maps:remove(ClientId, Ids)};
-                        _                 -> {ok, Ids}
-                    end,
-    {reply, Reply, State#state{ client_ids = Ids1 }};
-
-handle_call(list, _From, State = #state{client_ids = Ids}) ->
-    {reply, maps:to_list(Ids), State};
-
-handle_call(Msg, _From, State) ->
-    {stop, {unhandled_call, Msg}, State}.
-
-handle_cast(Msg, State) ->
-    {stop, {unhandled_cast, Msg}, State}.
-
-handle_info({'EXIT', _, {shutdown, closed}}, State) ->
-    {stop, {shutdown, closed}, State};
-
-handle_info({'DOWN', MRef, process, DownPid, _Reason},
-            State = #state{client_ids = Ids}) ->
-    Ids1 = maps:filter(fun (ClientId, {Pid, M})
-                           when Pid =:= DownPid, MRef =:= M ->
-                               rabbit_log_connection:warning(
-                                              "MQTT disconnect from ~p~n",
-                                              [ClientId]),
-                               false;
-                           (_, _) ->
-                               true
-                       end, Ids),
-    {noreply, State #state{ client_ids = Ids1 }}.
-
-terminate(_Reason, _State) ->
-    ok.
-
-code_change(_OldVsn, State, _Extra) ->
-    {ok, State}.
+%%----------------------------------------------------------------------------
+-spec run_ra_command(term()) -> term() | {error, term()}.
+run_ra_command(RaCommand) ->
+    NodeId = mqtt_node:node_id(),
+    case ra:process_command(NodeId, RaCommand) of
```

Comment:
given that the behaviour here is "last write wins" do we need to do a synchronous call here?  can we use pipeline command instead and handle the success notification?

Reply:
@kjnilsson and I agreed that it's not worth doing unless we have evidence of meaningful latency spikes for real world workloads.